scripts/train_linker.py

```diff
@@ -7,60 +7,103 @@
 import argparse
 from tqdm import tqdm
 import json
+import numpy as np
+from collections import defaultdict
 
 def read_file(filename, limit):
+        data = []
         x = []
         y = []
         with open(filename) as f:
             for line in tqdm(f, total=limit):
                 d = json.loads(line)
+                data.append(d)
                 x.append(Linker.featurizer(d))
                 y.append(d['label'])
                 if len(x) >= limit:
                     break
-        return x, y
+        return data, x, y
 
-def main(data_path: str):
+def entity_level_eval(data, all_features, classifier):
+    k_list = [1, 3, 5, 10]
+    classifier_correct_predictions = defaultdict(int)
+    classifier_total_predictions = defaultdict(int)
+    candidate_features = []
+    gold_index = -1
+    for i, d in tqdm(enumerate(data)):
+        candidate_index = d['candidate_index']
+        if candidate_index == 0:
```

Comment:
Can you collapse these if statements into a single block, so the code is less indented? It makes it easier to read.

Additionally, if you refactor to loop over candidate features from the start, you won't have to check its length here.